lib/realtime/vehicles.ex

```diff
@@ -44,48 +52,87 @@ defmodule Realtime.Vehicles do
     vehicles_and_ghosts = ghosts ++ ungrouped_vehicles
 
-    incoming_from_another_route =
-      incoming_from_another_route(incoming_trips, vehicles_and_ghosts)
-      |> Enum.map(fn {route_id, vehicles_and_ghosts} ->
-        {route_id,
-         sort_incoming_vehicles_and_ghosts(route_id, vehicles_and_ghosts, incoming_trips)}
-      end)
-      |> Map.new()
+    incoming_from_another_route = incoming_from_another_route(incoming_trips, vehicles_and_ghosts)
 
     vehicles_and_ghosts
     |> Enum.filter(fn vehicle_or_ghost -> vehicle_or_ghost.route_id != nil end)
     |> Enum.group_by(fn vehicle_or_ghost -> vehicle_or_ghost.route_id end)
-    |> Map.merge(incoming_from_another_route, fn _route_id, on_route, incoming ->
-      on_route ++ incoming
+    |> Map.merge(incoming_from_another_route, fn route_id, on_route, interlining ->
+      {pulling_out, not_pulling_out} =
+        Enum.split_with(on_route, &(&1.route_status == :pulling_out))
+
+      not_pulling_out ++
+        sort_incoming_vehicles_and_ghosts(
+          route_id,
+          pulling_out,
+          interlining,
+          incoming_trips,
+          date_by_block_id
+        )
     end)
   end
 
-  @spec sort_incoming_vehicles_and_ghosts(Route.id(), [Vehicle.t() | Ghost.t()], [Trip.t()]) :: [
+  @spec sort_incoming_vehicles_and_ghosts(
+          Route.id(),
+          [Vehicle.t() | Ghost.t()],
+          [Vehicle.t() | Ghost.t()],
+          [Trip.t()],
+          %{Block.id() => Date.t()}
+        ) :: [
           Vehicle.t() | Ghost.t()
         ]
-  defp sort_incoming_vehicles_and_ghosts(route_id, vehicles_and_ghosts, incoming_trips) do
-    vehicles_and_ghosts
-    |> Enum.sort_by(fn vehicle_or_ghost ->
-      incoming_trip =
-        incoming_trips
-        |> Enum.find(fn trip ->
-          trip.block_id == vehicle_or_ghost.block_id && trip.route_id == route_id
-        end)
+  defp sort_incoming_vehicles_and_ghosts(
+         route_id,
+         pulling_out_vehicles_and_ghosts,
+         interlining_vehicles_and_ghosts,
+         incoming_trips,
+         date_by_block_id
+       ) do
+    interlining_start_timestamps =
+      Map.new(
+        interlining_vehicles_and_ghosts,
+        fn vehicle_or_ghost ->
+          incoming_trip =
+            incoming_trips
+            |> Enum.find(fn trip ->
+              trip.block_id == vehicle_or_ghost.block_id && trip.route_id == route_id
+            end)
 
-      incoming_trip.start_time
-    end)
+          block_date = Map.fetch!(date_by_block_id, incoming_trip.block_id)
+
+          incoming_trip_start_timestamp =
+            Util.Time.timestamp_for_time_of_day(incoming_trip.start_time, block_date)
+
+          {vehicle_or_ghost, incoming_trip_start_timestamp}
+        end
+      )
+
+    pulling_out_start_timestamps =
+      Map.new(pulling_out_vehicles_and_ghosts, fn vehicle_or_ghost ->
+        {vehicle_or_ghost, vehicle_or_ghost.layover_departure_time}
```

Comment:
It feels a little weird to have this big object as a map key. 3 options:
1. It's not a big deal, leave it.
2. Use the vehicle id as a key instead.
3. Instead of a map, make it a list with elements [{vehicle, time}], then sort the list, then map over it to drop the times.

Reply:
I'm inclined to leave it, I don't think it's a big deal to use a whole object as a hash key when it's convenient.